[benchpress] add tee_stdout and tee_stderr to job

vmagro opened the pull request:
It was very difficult to get the output of benchpress jobs' subprocesses
This adds the capability to add `tee_stdout` and `tee_stderr` options to
a job config and have benchpress log `stdout` and `stderr` to it's
`stdout`
These options default to False, they can be set to True to log to
`stdout` or to a string indicating the path to a file to write to.

meteorfox commented:
Just trying to help, not trying to be condescending or anything.

Here's an example I put together for configuring subprocess and logging to output either stdout and stderr to console and/or log file using logging levels. 

Hopefully this can help you get it working with subprocess:
https://gist.github.com/meteorfox/4233bb0f5e0aa1915b48c8854f3142b1

If you want to keep similar semantics to `subprocess.communicate()` then you could use the newer API `subprocess.run()`, instead of `subprocess.Popen()` that takes care of timeouts and exposes the same API argument as Popen(), you could also set check=False, so you can do the check yourself for errors.

https://docs.python.org/3/library/subprocess.html#subprocess.run

Let me know if you have any questions, or if you want me to take at stab at it.

vmagro commented:
@meteorfox there was some reason that I used `Popen` instead of `run`, but I can't remember what it was now. 
I do like your comment about just passing file descriptors instead of reading into a string, I'll look into that

greenscientist commented on benchpress/benchpress/lib/job.py:
```diff
@@ -104,6 +107,23 @@ def run(self):
+                if tee != sys.stdout:
+                    tee.close()
+            if self.tee_stderr:
+                tee = sys.stdout
```
don't you want sys.stderr here?

greenscientist commented on benchpress/benchpress/lib/job.py:
```diff
@@ -104,6 +107,23 @@ def run(self):
+                    tee = open(self.tee_stderr, 'w')
+                for line in stderr.splitlines():
+                    tee.write('stderr: {}\n'.format(line))
+                if tee != sys.stdout:
```
and here?

vmagro commented:
Amended the commit, simplified to be just one option `tee_output` where both `stdout` and `stderr` go to the same location, added some comments describing what the code is doing
